### ISSUE TITLE
Calling useTranslation may cause runtime error (Next 12.2 on React 18.2 with Streaming SSR)

### ISSUE THREAD

**Issue by AppleNectar** [i18next/next-i18next#1924]
## 🐛 Bug Report

Please forgive me for listing the issue in this repository, although it is possible that this problem is caused by Next or React.

The following runtime error occurs when `useTranslation` is called on a component doing Streaming SSR in Next.js 12.2.

```
Unhandled Runtime Error

Error: This Suspense boundary received an update before it finished hydrating. This caused the boundary to switch to client rendering. The usual way to fix this is to wrap the original update in startTransition.
```

If `useTranslation` is not called, this problem does not occur.  
However, this only occurs with developer builds and not with production builds.  
Also, if the strict mode of React is turned off, it will not occur even in developer builds.

## To Reproduce

[CodeSandbox](https://codesandbox.io/s/next-js-hydration-issue-omf1cf?file=/pages/index.tsx)

```tsx
import { GetServerSideProps, NextPage } from "next";
import { SSRConfig, useTranslation } from "next-i18next";
import { serverSideTranslations } from "next-i18next/serverSideTranslations";
import { Suspense } from "react";

type Props = SSRConfig;

let data: string;

const SampleComponent = (): JSX.Element => {
  if (!data) {
    throw new Promise(async (resolve) => {
      // The reason for the 200 ms stop is that this problem does not occur when the response time is around 100 ms.
      await new Promise((resolve) => setTimeout(resolve, 200));

      data = "resolved!";

      resolve(true);
    });
  }

  return <div>{data}</div>;
};

const IndexPage: NextPage<Props> = () => {
  // Runtime Error occurs when calling the hook of next-i18next (react-i18next)
  const { t } = useTranslation(["common"]);

  return (
    <div>
      <Suspense fallback={<div>Loading...</div>}>
        <SampleComponent />
      </Suspense>
    </div>
  );
};

export const getServerSideProps: GetServerSideProps<Props> = async (
  context
) => {
  return {
    props: {
      ...(await serverSideTranslations(context.locale ?? "en"))
    }
  };
};

export default IndexPage;
```

## Expected behavior

Runtime error (`This Suspense boundary received an update before it finished hydrating.`) is not occurred.

## Your Environment

- *runtime version*: node v16.15.0
- *React version*: 18.2.0
- *Next.js version*: 12.2.5
- *i18next version*: 21.8.14
- *next-i18next version*: 11.3.0
- *os*: Ubuntu-20.04

**Comment by adrai** [comment]
Currently afk...
Can you try to call useTranslation inside (under) Suspese?

**Comment by adrai** [comment]
btw: by default useSuspense is set to false in next-i18next

**Comment by AppleNectar** [comment]
@adrai 
> Can you try to call useTranslation inside (under) Suspese?

Sorry, I omitted to mention this.
The problem does not occur when called in a suspense.
```tsx
const SampleComponent = (): JSX.Element => {
  if (!data) {
    throw new Promise(async (resolve) => {
      // The reason for the 200 ms stop is that this problem does not occur when the response time is around 100 ms.
      await new Promise((resolve) => setTimeout(resolve, 200));

      data = "resolved!";

      resolve(true);
    });
  }

  const { t } = useTranslation(["common"]);

  return <div>{data}</div>;
};
```
But I want to use it outside of Suspense ...

**Comment by adrai** [comment]
In a normal react-i18next setup useTranslation always needs to be inside a suspense.

**Comment by AppleNectar** [comment]
@adrai 
I understood the useTranslation hook will trigger a Suspense if not ready.
However, the results did not change when we tried each of the following.

* Wrap page component with Suspense in `_app.tsx` ( useSuspense: true )
```tsx
import { appWithTranslation } from "next-i18next";
import { Suspense } from "react";
import type { AppProps } from "next/app";

function MyApp({ Component, pageProps }: AppProps) {
  return (
    <Suspense fallback={<div>Loading...</div>}>
      <Component {...pageProps} />
    </Suspense>
  );
}

export default appWithTranslation(MyApp);
```
* Disable the useSuspense option of react-i18next
```tsx
const { t } = useTranslation(["common"], { useSuspense: false });
```

Do you have any solutions?

**Comment by adrai** [comment]
The problem is "await new Promise((resolve) => setTimeout(resolve, 200));" is "detected" in useTranslation

**Comment by adrai** [comment]
Maybe @jamuhl knows more, but if you're throwing promises, you also need to make sure useTranslation is inside a Suspense component...

**Comment by jamuhl** [comment]
I think it's really just that you're using `useTranslation` outside any Suspense Component -> Your Suspense Component is inside the component calling useTranslation (Suspense does not wrap it)

**Comment by AppleNectar** [comment]
@jamuhl 
In short, does it mean that the problem does not occur if either of the following is true?

If the `useSuspense` option is set to `false`.
-> `useTranslation`, but you need to refer to `ready`, the return value of `useTranslation`, to determine if the resource has been loaded or not.

If the `useSuspense` option is set to `true`.
-> Wrap the component calling `useTranslation` with `<Suspens>`.

**Comment by adrai** [comment]
https://react.i18next.com/latest/usetranslation-hook#not-using-suspense

**Comment by AppleNectar** [comment]
@adrai 
I have already checked your documentation.

As mentioned in several previous  [reply](https://github.com/i18next/next-i18next/issues/1924#issuecomment-1212953984), setting `useSuspense` to `false` does not improve it.

**Comment by adrai** [comment]
it does not, because you a throwing a promise and useTranslation detects it (this is a React behaviour)

**Comment by adrai** [comment]
tldr;
- next-i18next sets useSuspense to false by default
- throwing a promise and having useTranslation outside of Suspense is causing this issue

advice: call useTranslation inside Suspense

if someone knows of a better solution with the new steaming ssr, provide a PR

**Comment by AppleNectar** [comment]
> useTranslation detects it

I don't know what this means.

I recognize that when a component throws a Promise, React behaves as follows.

1. Rendering of the thrown component is suspended
2. Nearest `<Suspense>` component catches it
3. Rendering `fallback` components
4. If the Promise is resolved, the suspended component is rendered instead of the fallback component

It is strange that the problem occurs even when `useTranslation` is outside of the `Suspense` component and `useSuspense` option is set to `false` so that `useTranslation` does not throw a Promise.

In any case, it seems that the problem cannot be solved immediately.
Thank you for your reply.

**Comment by adrai** [comment]
I'm afk, can you check if this line is reached? https://github.com/i18next/react-i18next/blob/a4fd6ef9de9758a1f0d790d533a8864b38b8d7d8/src/useTranslation.js#L118

**Comment by jamuhl** [comment]
if set to useSuspense to false the useTranslation does not throw...https://github.com/i18next/react-i18next/blob/master/src/useTranslation.js#L115 <<--- proof

It's the nearest Suspense higher in the tree...so yes Suspense must be a component higher - not inside the render of the component using useTranslation

**Comment by AppleNectar** [comment]
@adrai 
If `useSuspense` is `false` (In other words, the default setting for next-i18next.), the line is never reached (i.e., no Promise is thrown).
Of course, Promise is thrown by setting `useSuspense` to `true` in the `useTranslation` argument.

@jamuhl 
>It's the nearest Suspense higher in the tree...so yes Suspense must be a component higher - not inside the render of the component using useTranslation

Yes, I understand that.
Next.js uses the component exported from `pages/_app.tsx` as the top-level component, and can render components like the one shown in this issue (under the pages directory) within it.
So I wrapped the `IndexPage` component shown in this issue in `_app.tsx` with `Suspense`.
But that was not the solution.

* pages/app.tsx
```tsx
import type { AppProps } from "next/app";
import { appWithTranslation } from "next-i18next";
import { Suspense } from "react";

function MyApp({ Component, pageProps }: AppProps) {
  return (
    <Suspense fallback={<div>Loading...</div>}>
      <Component {...pageProps} />
    </Suspense>
  );
}

export default appWithTranslation(MyApp);
```

* pages/index.tsx
```tsx
import { GetServerSideProps, NextPage } from "next";
import { SSRConfig, useTranslation } from "next-i18next";
import { serverSideTranslations } from "next-i18next/serverSideTranslations";
import { Suspense } from "react";

type Props = SSRConfig;

let data: string;

const SampleComponent = (): JSX.Element => {
  if (!data) {
    throw new Promise(async (resolve) => {
      await new Promise((resolve) => setTimeout(resolve, 200));

      data = "resolved!";

      resolve(true);
    });
  }

  return <div>{data}</div>;
};

const IndexPage: NextPage<Props> = () => {
  // Isn't the Promise thrown here caught by the Suspense of the MyApp component of _app.tsx ?
  const { t } = useTranslation(["common"], { useSuspense: true });

  return (
    <div>
      <Suspense fallback={<div>Loading...</div>}>
        <SampleComponent />
      </Suspense>
    </div>
  );
};

export const getServerSideProps: GetServerSideProps<Props> = async (
  context
) => {
  return {
    props: {
      ...(await serverSideTranslations(context.locale ?? "en"))
    }
  };
};

export default IndexPage;
```

**Comment by adrai** [comment]
> If `useSuspense` is `false` (In other words, the default setting for next-i18next.), the line is never reached (i.e., no Promise is thrown).

ok, so seems not really related to that.... hmmmm

**Comment by adrai** [comment]
fyi: it seems not really related to useTranslation directly...

Try:


```js
const IndexPage: NextPage<Props> = () => {
  // Runtime Error occurs when calling the hook of next-i18next (react-i18next)
  // const { t } = useTranslation('second-page')
  const [value, setValue] = useState()
  useEffect(() => {
    setTimeout(() => {
      setValue(1)
    }, 10)
  }, [])

  return (
    <div>
      <Suspense fallback={<div>Loading...</div>}>
        <SampleComponent />
      </Suspense>
    </div>
  );
};
```


<img width="982" alt="image" src="https://user-images.githubusercontent.com/1086194/184384503-746ac8dc-dfd7-4dd7-8db4-da30cf6188f2.png">

**Comment by AppleNectar** [comment]
@adrai 
@jamuhl 
Oh my god...

I apologize for the lack of verification...
It seems to be a problem on the Next.js side, so I will consider putting an issue in the repository there.

Thank you for responding so quickly.
Thank you so much.
